contextlib2/contextlib.pyi

```diff
@@ -0,0 +1,122 @@
+# Type hints copied from the typeshed project under the Apache License 2.0
+# https://github.com/python/typeshed/blob/64c85cdd449ccaff90b546676220c9ecfa6e697f/LICENSE
+
+import sys
+from types import TracebackType
+from typing import (
+    IO,
+    Any,
+    AsyncContextManager,
+    AsyncIterator,
+    Awaitable,
+    Callable,
+    ContextManager,
+    Iterator,
+    Optional,
+    Type,
+    TypeVar,
+    overload,
+)
+from typing_extensions import ParamSpec, Protocol
+
+AbstractContextManager = ContextManager
+if sys.version_info >= (3, 7):
+    AbstractAsyncContextManager = AsyncContextManager
+
+_T = TypeVar("_T")
+_T_co = TypeVar("_T_co", covariant=True)
+_T_io = TypeVar("_T_io", bound=Optional[IO[str]])
+_F = TypeVar("_F", bound=Callable[..., Any])
+_P = ParamSpec("_P")
+
+_ExitFunc = Callable[[Optional[Type[BaseException]], Optional[BaseException], Optional[TracebackType]], bool]
+_CM_EF = TypeVar("_CM_EF", ContextManager[Any], _ExitFunc)
+
+class _GeneratorContextManager(ContextManager[_T_co]):
+    def __call__(self, func: _F) -> _F: ...
+
+# type ignore to deal with incomplete ParamSpec support in mypy
+def contextmanager(func: Callable[_P, Iterator[_T]]) -> Callable[_P, _GeneratorContextManager[_T]]: ...  # type: ignore
+
+if sys.version_info >= (3, 7):
```

Comment:
Isn't this always defined in contextlib2?

Reply:
You could swap them out for `if True:` to minimize the diff noise?

Reply:
Ah, true. And I like that suggestion for minimising the diff.

Reply:
The ABC will be a bit troublesome as you might need to import them from typing_extensions?

Reply:
I wasn't sure about that, as the typing module imports were unconditional in the typeshed version as well - it only guarded their aliasing to contextlib API entries.

I figured either mypy (et al) must be doing some shenanigans behind the scenes to let that work even on older Python versions, or else it needed to be fixed in typeshed first, and then contextlib2 could inherit that fix.

Reply:
Ah you're right, it's already in 3.6 https://docs.python.org/3.6/library/typing.html#typing.AsyncContextManager

Reply:
Heh, we both found the same info in about the same time, but picked different subthreads to update first :)